From 0bf6f31afe34a1f54eb74a8d4a59d5a4924d6bed Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Tue, 14 Oct 2025 07:29:42 +0200
Subject: Remove special treatment for embedded hyperlinks

We don't want the links as an extra element in the a11y hierarchy to allow us
to tell the screen reader that the link will open in a browser and with which
URL. The embedded element is not transparent to us and we can't set any
properties on it. So we remove it altogether.

Change-Id: I19c4b5e17250cf9933ef635f1370f394058676f3
---
 src/quick/accessible/qaccessiblequickitem.cpp | 52 +------------------
 1 file changed, 1 insertion(+), 51 deletions(-)

diff --git x/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp y/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
index c33700d4e2ccae7d062adc1e845847528d8e40af..79f0fd81e975aa6c30c312ec7ff00082358061d1 100644
--- x/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
+++ y/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
@@ -235,16 +235,7 @@ QWindow *QAccessibleQuickItem::window() const
     return window;
 }
 
-int QAccessibleQuickItem::childCount() const
-{
-    // see comment in QAccessibleQuickItem::child() as to why we do this
-    int cc = 0;
-    if (QQuickText *textItem = qobject_cast<QQuickText*>(item())) {
-        cc = QQuickTextPrivate::get(textItem)->getLinks().size();
-    }
-    cc += childItems().size();
-    return cc;
-}
+int QAccessibleQuickItem::childCount() const { return childItems().size(); }
 
 QRect QAccessibleQuickItem::rect() const
 {
@@ -277,17 +268,6 @@ QAccessibleInterface *QAccessibleQuickItem::childAt(int x, int y) const
             return nullptr;
     }
 
-    // special case for text interfaces
-    if (QQuickText *textItem = qobject_cast<QQuickText*>(item())) {
-        const auto hyperLinkChildCount = QQuickTextPrivate::get(textItem)->getLinks().size();
-        for (auto i = 0; i < hyperLinkChildCount; i++) {
-            QAccessibleInterface *iface = child(i);
-            if (iface->rect().contains(x,y)) {
-                return iface;
-            }
-        }
-    }
-
     // general item hit test
     const QList<QQuickItem*> kids = accessibleUnignoredChildren(item(), true);
     for (int i = kids.size() - 1; i >= 0; --i) {
@@ -364,22 +344,6 @@ QAccessibleInterface *QAccessibleQuickItem::child(int index) const
     if (index < 0)
         return nullptr;
 
-
-    if (QQuickText *textItem = qobject_cast<QQuickText*>(item())) {
-        const int hyperLinkChildCount = QQuickTextPrivate::get(textItem)->getLinks().size();
-        if (index < hyperLinkChildCount) {
-            auto it = m_childToId.constFind(index);
-            if (it != m_childToId.constEnd())
-                return QAccessible::accessibleInterface(it.value());
-
-            QAccessibleHyperlink *iface = new QAccessibleHyperlink(item(), index);
-            QAccessible::Id id = QAccessible::registerAccessibleInterface(iface);
-            m_childToId.insert(index, id);
-            return iface;
-        }
-        index -= hyperLinkChildCount;
-    }
-
     QList<QQuickItem *> children = childItems();
     if (index < children.size()) {
         QQuickItem *child = children.at(index);
@@ -390,22 +354,8 @@ QAccessibleInterface *QAccessibleQuickItem::child(int index) const
 
 int QAccessibleQuickItem::indexOfChild(const QAccessibleInterface *iface) const
 {
-    int hyperLinkChildCount = 0;
-    if (QQuickText *textItem = qobject_cast<QQuickText*>(item())) {
-        hyperLinkChildCount = QQuickTextPrivate::get(textItem)->getLinks().size();
-        if (QAccessibleHyperlinkInterface *hyperLinkIface = const_cast<QAccessibleInterface *>(iface)->hyperlinkInterface()) {
-            // ### assumes that there is only one subclass implementing QAccessibleHyperlinkInterface
-            // Alternatively, we could simply iterate with child() and do a linear search for it
-            QAccessibleHyperlink *hyperLink = static_cast<QAccessibleHyperlink*>(hyperLinkIface);
-            if (hyperLink->textItem() == static_cast<QQuickText*>(item())) {
-                return hyperLink->linkIndex;
-            }
-        }
-    }
     QList<QQuickItem*> kids = childItems();
     int idx = kids.indexOf(static_cast<QQuickItem*>(iface->object()));
-    if (idx >= 0)
-        idx += hyperLinkChildCount;
     return idx;
 }
 
